src/data_store/data_store_conduit.cpp

```diff
@@ -28,26 +28,47 @@
 #include "lbann/data_store/data_store_conduit.hpp"
 
 #include "lbann/data_readers/data_reader_jag_conduit.hpp"
+#include "lbann/data_readers/data_reader_image.hpp"
 #include "lbann/utils/exception.hpp"
 #include "lbann/utils/options.hpp"
 #include "lbann/utils/timer.hpp"
 #include <unordered_set>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <unistd.h>
 
 namespace lbann {
 
+// Macro to throw an LBANN exception
+#undef LBANN_ERROR
+#define LBANN_ERROR(message)                                    \
+  do {                                                          \
+    std::stringstream ss_LBANN_ERROR;                           \
+    ss_LBANN_ERROR << "LBANN error ";                           \
+    const int rank_LBANN_ERROR = lbann::get_rank_in_world();    \
+    if (rank_LBANN_ERROR >= 0) {                                \
+      ss_LBANN_ERROR << "on rank " << rank_LBANN_ERROR << " ";  \
+    }                                                           \
+    ss_LBANN_ERROR << "(" << __FILE__ << ":" << __LINE__ << ")" \
+                     << ": " << (message);                      \
+    if (errno) {                                                \
+      ss_LBANN_ERROR << "\nerrno: " << errno << " msg: "        \
+                     << strerror(errno);                        \
+    }                                                           \
+    if (m_output) {                                             \
+      m_output << "ERROR: " << ss_LBANN_ERROR.str()             \
+               << std::endl;                                    \
+      m_output.close();                                         \
+    }                                                           \
+    throw lbann::exception(ss_LBANN_ERROR.str());               \
+  } while (0)
+
```

Comment:
m_output if a private variable (filestream) in data_store_conduit, so it wouldn't work to move it to exception.hpp  It's important to have the filestream closed if an exception is thrown, so I can tell what when wrong. Eventually m_output (and associated writes thereto) will go away; at that time the private LBANN_EXCEPTION in data_store_conduit can also go away.

Reply:
Also: I think we should not add printing errno to the LBANN_ERROR in exceptions.hpp because:
"The value of errno is defined only after a call to a function for which it is explicitly stated to be set"
Hence, the value of errno may be meaningless for many places in the codebase where an lbann exception is thrown. I added it to the private LBANN_ERROR in data_store_conduit for temporary debugging, because the calls creating the shared memory segment will set it on error.